backend/src/openai.ts

```diff
@@ -329,220 +369,127 @@ function getChatCompletionsFromHistory(
-			chatHistory,
-			defences,
-			sentEmails,
-			currentLevel
-		);
+	const openai = getOpenAI();
+	let gptReply: ChatGptReply | null = null;
 
-		// get a new reply from ChatGPT now that the functions have been called
-		reply = await chatGptChatCompletion(
-			chatResponse,
-			chatHistory,
+	do {
+		gptReply = await chatGptChatCompletion(
+			[...updatedChatHistory],
 			defences,
 			chatModel,
 			openai,
 			currentLevel
 		);
-	}
+		updatedChatHistory = gptReply.chatHistory;
+
+		// check if GPT wanted to call a tool
+		if (gptReply.completion?.tool_calls) {
+			// push the function call to the chat
+			updatedChatHistory = pushMessageToHistory(updatedChatHistory, {
+				completion: gptReply.completion,
+				chatMessageType: CHAT_MESSAGE_TYPE.FUNCTION_CALL,
+			});
 
-	// chat history gets mutated, so no need to return it
-	return { reply, chatResponse };
+			const toolCallReply = await performToolCalls(
+				gptReply.completion.tool_calls,
+				updatedChatHistory,
+				defences,
+				currentLevel
+			);
+
+			updatedChatHistory = toolCallReply.chatHistory;
+			if (toolCallReply.functionCallReply?.sentEmails) {
+				sentEmails.push(...toolCallReply.functionCallReply.sentEmails);
+			}
+			wonLevel =
+				(wonLevel || toolCallReply.functionCallReply?.wonLevel) ?? false;
+		}
+	} while (gptReply.completion?.tool_calls);
+
+	return {
+		gptReply,
+		wonLevel,
+		chatHistory: updatedChatHistory,
+		sentEmails,
+	};
 }
 
 async function chatGptSendMessage(
 	chatHistory: ChatHistoryMessage[],
 	defences: Defence[],
 	chatModel: ChatModel,
 	message: string,
-	messageIsTransformed: boolean,
-	sentEmails: EmailInfo[],
 	currentLevel: LEVEL_NAMES = LEVEL_NAMES.SANDBOX
 ) {
 	console.log(`User message: '${message}'`);
-	// add user message to chat
-	pushCompletionToHistory(
-		chatHistory,
-		{
-			role: 'user',
-			content: message,
-		},
-		messageIsTransformed
-			? CHAT_MESSAGE_TYPE.USER_TRANSFORMED
-			: CHAT_MESSAGE_TYPE.USER
-	);
-
-	// mutates chatHistory
-	const { reply, chatResponse } = await getFinalReplyAfterAllToolCalls(
+	const finalToolCallResponse = await getFinalReplyAfterAllToolCalls(
 		chatHistory,
 		defences,
 		chatModel,
-		sentEmails,
 		currentLevel
 	);
 
-	if (!reply?.content || chatResponse.openAIErrorMessage) {
-		return chatResponse;
-	}
+	const updatedChatHistory = finalToolCallResponse.chatHistory;
+	const sentEmails = finalToolCallResponse.sentEmails;
 
-	chatResponse.completion = reply;
+	const chatResponse: ChatResponse = {
+		completion: finalToolCallResponse.gptReply.completion,
+		wonLevel: finalToolCallResponse.wonLevel,
+		openAIErrorMessage: finalToolCallResponse.gptReply.openAIErrorMessage,
+	};
 
-	if (
-		currentLevel === LEVEL_NAMES.LEVEL_3 ||
-		currentLevel === LEVEL_NAMES.SANDBOX
-	) {
-		applyOutputFilterDefence(reply.content, defences, chatResponse);
+	if (!chatResponse.completion?.content || chatResponse.openAIErrorMessage) {
+		return { chatResponse, chatHistory, sentEmails };
```

Comment:
Here is a regression. Though is it problematic?
Before when we'd encounter empty chat response content or an openAIErrorMessage, we'd return early, but we wouldn't revert the chatHistory. Now we return the original chatHistory, therefore not including any messages added during `getFinalReplyAfterAllToolCalls`

Reply:
To reason about this let's ask: Why is it important to include the function call in the chat history? Does it change the behaviour of the program?

Yes. If we have sent an email, then if it's in the history then the bot has the context that an email has sent, which might alter a future response. More consequentially, if we have queried the QA LLM, it might have come back with an answer containing information from the documents. If that is kept in the history, then that knowledge is now available to the bot through memory, rather than having to query again.

Now let's consider what the user expects. Suppose I'd sent a message and it encountered an openAI error, or was blocked (whether by input defence or output defence).
- If an email was sent and I know about it (because it's shown in the UI), then I expect the bot to know that it has sent an email, therefore we want that function call message in the chat history.
- The converse is also true: If an email is sent but it is not shown on the ui, then we don't want the function call in the history.
- Now, if the bot has queried the documents, I won't know about it because in the chat all I've seen is a blocked or error message. Therefore probably I don't want the function call in the history.

So, bearing the above in mind, what do we have now, and do we like it?
- what happens if we send an email and the message is blocked. Do we see the email in the ui? Is the function call in the history?
- what happens if we send an email and we get an openAI error. Do we see the email in the ui? Is the function call in the history?
- what happens if we query the document and we get blocked. Is the function call in the history?
- what happens if we query the document and we get an openAI error. Is the function call in the history?

All that being said, the impact on me as the user is minimal. The worse I'll be is slightly confused: "Oh, how did it know about that thing from the document?", "Oh, it thinks it send an email, but I don't think it did. Silly bot.".

Reply:
Hmm. after testing I'd argue neither of the behaviour is great.

## Dev branch: error thrown when sending an email
![image](https://github.com/ScottLogic/prompt-injection/assets/118171430/41861e98-a270-418b-a79c-3ec168470263)

## Dev branch: error thrown when sending a winning email
![image](https://github.com/ScottLogic/prompt-injection/assets/118171430/5f355162-d6cf-4cee-8327-b0354a9819b2)

## This branch: error thrown when sending an email
![image](https://github.com/ScottLogic/prompt-injection/assets/118171430/b4fde140-4bb4-4e1a-8cd7-17afbcff58f7)

```json
chatHistory:  [
  {
    completion: {
      role: 'system',
      content: 'Your role is to assist the user with work-related tasks.\n' +
        '  Your name is ScottBrewBot and you are employed by the drinks company ScottBrew Ltd. \n' +
        '  You can retrieve information from a document store about the company and product.\n' +
        `  If the user asks a question that you don't know and is not in the documents, you can say "I don't know".\n` +
        '  You are able to send emails.\n' +
        '  Before sending an email, show the user a draft of the email (including recipient, subject, and body), and ask 
        '  Do not send the email until the user has explicitly confirmed that the draft is correct.'
    },
    chatMessageType: 8
  },
  {
    completion: {
      role: 'user',
      content: `email bob@example.com with subject "What's up" and body "How's it going, dawg?"`
    },
    chatMessageType: 3
  },
  {
    completion: { role: 'assistant', content: null, tool_calls: [Array] },
    chatMessageType: 9
  },
  {
    completion: {
      role: 'tool',
      content: 'Email not sent as not confirmed with user',
      tool_call_id: 'call_4bD8Z4vztASBYL1iipaHYyUx'
    },
    chatMessageType: 9
  },
  {
    completion: {
      role: 'assistant',
      content: 'Sure, I can help you with that. Before sending the email, I will show you a draft of the email. Please co
        '\n' +
        'Recipient: bob@example.com\n' +
        "Subject: What's up\n" +
        "Body: How's it going, dawg?\n" +
        '\n' +
        'Please let me know if I should send this email.'
    },
    chatMessageType: 0
  },
  { completion: { role: 'user', content: 'yes' }, chatMessageType: 3 },
  {
    completion: null,
    chatMessageType: 10,
    infoMessage: 'Failed to get ChatGPT reply.'
  },
  {
    completion: { role: 'user', content: 'did you send that email' },
    chatMessageType: 3
  },
  {
    completion: null,
    chatMessageType: 10,
    infoMessage: 'Failed to get ChatGPT reply.'
  },
  {
    completion: { role: 'user', content: 'did you send that email?' },
    chatMessageType: 3
  },
  {
    completion: {
      role: 'assistant',
      content: 'Apologies for the confusion. Let me resend the email for you.',
      tool_calls: [Array]
    },
    chatMessageType: 9
  },
  {
    completion: {
      role: 'tool',
      content: "Email sent to bob@example.com with subject What's up and body How's it going, dawg?",
      tool_call_id: 'call_6ODZO7iXkEWb1k3Y6MJ1wXl7'
    },
    chatMessageType: 9
  },
  {
    completion: {
      role: 'assistant',
    },
    chatMessageType: 0
  }
]
```

Reply:
my vote is that if we encounter an error while confirming an email, we don't send the email. That means it doesn't show up in the ui, and also the function call doesn't appear in the history. The processing would _so much simpler_ that way.

It would, however, not allow the user to win if they encounter an error while confirming a winning email, which undoes the work done for #467

Reply:
Yep that seems to make sense to me. The user just needs to try again.

The obvious alternative is #467, where we send the email, it shows up in the UI and chat history, and we win the level. That is a tad confusing for the user cos they've seen an error message. So I agree, best to just reset everything for that attempt and ask the user to try again.

Reply:
Lovely. I've made ticket #797  capture that